app/javascript/components/visualization/controls/ScatterPlotLegend.jsx

```diff
@@ -58,6 +59,43 @@ function LegendEntry({
     setShowColorPicker(false)
   }
 
+  /**
+   * Handle mouse enter events used for hovering by
+   * wrapping setting the active label in a debounce, this
+   * will delay the call to active trace label by 1000ms and
+   * can be canceled
+   */
+  const debouncedHandleMouseEnter = debounce(() => {
+    setActiveTraceLabel(label)
+  }, 1000)
```

Comment:
I don't think an array will give us much benefit here.

> Execute those arrayed handlers one at time, the next starting only after the previous completed

The issue is once Plotly starts to redraw the plot there is a lag. The way debounce is being used here, is to try to prevent the redraw even starting if the user is continuing to move their mouse. 

> Cancel _all handlers except the most recently pushed_ when a new handler is pushed

Both `onMouseEnter` and `onMouseLeave` are canceling the previous call to `setActiveTraceLabel` which, leaves only the latest call to `setActiveTraceLabel` to actually execute.

> Within each handler, don't wait -- executive immediately after the previously-pushed handler completes

The wait/delay is necessary to give time for the cancelation to occur. Otherwise, without a delay, the call to `setActiveTraceLabels` would occur immediately and once that's started the plot is already in the process of going to redraw which is what is causing the lag.


> ...option might be to pass numLabels into the top-level handlers, and only use debounce if it exceeds, say, 500 labels. 

I would very much like to avoid having different UX for different studies i.e. different delay times for the 'common case' studies and the 'extremely many label' studies because this would require users to understand that the same functionality will occur differently in different studies and could be confusing i.e. hovering in one study produces immediate results but hovering in others the results have a delay. If we go that route I think we'd need a way to let the users know that this is expected.


I've played around with the timing of the delay, and currently 600ms seems to look not laggy in the 'common case' study and provides some benefit to the lag of a 'extremely many labels' study. See how the 600ms feels for you.

Reply:
> avoid having different UX for different studies i.e. different delay times for the 'common case' studies and the 'extremely many label' studies because this would require users to understand that the same functionality will occur differently in different studies and could be confusing i.e. hovering in one study produces immediate results but hovering in others the results have a delay. If we go that route I think we'd need a way to let the users know that this is expected.

Hovering will be delayed in large studies regardless of what we do here, because it takes a while for Plotly to render it.  Users intuit that large datasets are slower, so I don't think we need to let users know that -- such rationale suggests we'd need to inform users even with the 600-ms wait.

We also have many cases of contextual optimizations as I propose here.  We don't let users know about any of them.  I'm unaware of any cases where we've consciously slowed down a common case to speed up an edge case.  

So the rationale for slowing down the common case seems odd to me, especially when we can avoid that by a seemingly straightforward condition.